---
title: "WCS-SD"
author: "Jordan Lee"
date: "3/6/2020"
output: html_document
---

# Purpose

Notebook will showcase the comparisons we are interested in presenting at the West Coast Symposium in San Diego 2020.

# Analysis Preprocessing

```{r, cache=TRUE}
# Disclaimer: hardcoded names of patients in flat_matrix
# hard coded condition vector
####################
file_bulk1 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7319_B2_COUNTmatrix.txt'
file_bulk2 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7320_B2_COUNTmatrix.txt'
file_bulkGT1 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7319_GT_COUNTmatrix.txt'
file_bulkGT2 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7320_GT_COUNTmatrix.txt'
file_bulkNORM1 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7319_N_COUNTmatrix.txt'
file_bulkNORM2 <-'/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/01_process-bulkrna/07_ct_matrices/SCC_7320_N_COUNTmatrix.txt'

countsMECH_1_matrix_dir <- '/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/02_process-scrna/7319_Mech/'
countsENZ_1_matrix_dir <- '/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/02_process-scrna/7319_Enz/'
countsMECH_2_matrix_dir <- '/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/02_process-scrna/7320_Mech/'
countsENZ_2_matrix_dir <- '/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/02_process-scrna/7320_Enz/'
THRES = 0.01
output_sortedGenes = '/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/03_compare-bulk_scrna/scrna2bulk/geneOrdered_results.csv'

bulk_base_names <- c('bulk_7319', 'bulk_7320')
bulkgt_base_names <- c('bulkGT_7319', 'bulkGT_7320')
bulknorm_base_names <- c('bulk_norm_7319', 'bulk_norm_7320')
scmech_base_names <- c('sc_mech_7319', 'sc_mech_7320')
scenz_base_names <- c('sc_enz_7319', 'sc_enz_7320')
####################
```

```{r, message=FALSE}
library(Seurat)
library(dplyr)
library(Matrix)
library(gdata)
library(DESeq2)
library(ggplot2)
library(pheatmap)
library(vsn)
library(RColorBrewer)
library(readr)
library(tibble)
library(tidyverse)
library(ggrepel)
```

# Create combined matrices of mech and enz 

Read in files and create pseudo-bulk (take the sum of raw gene counts across all cell barcodes and report that as the pseudo-bulk gene count). Preprocessing steps shown in `.Rmd`

```{r, cache=TRUE}
###############################
# Create combined count matrices
###############################
# Read in files
counts_mech1 <- Read10X(data.dir = countsMECH_1_matrix_dir, gene.column = 1)
counts_enz1 <- Read10X(data.dir = countsENZ_1_matrix_dir, gene.column = 1)
counts_mech2 <- Read10X(data.dir = countsMECH_2_matrix_dir,gene.column = 1)
counts_enz2 <- Read10X(data.dir = countsENZ_2_matrix_dir, gene.column = 1)
# convert to dense matrix
counts_mech1 <- as.matrix(counts_mech1) 
counts_enz1 <- as.matrix(counts_enz1) 
counts_mech2 <- as.matrix(counts_mech2) 
counts_enz2 <- as.matrix(counts_enz2) 
######## Subset for testing ###########
# counts_mech1 <- counts_mech1[256:260, 1:3]
# counts_mech1 <- counts_mech1[493:498, 1:3]
# counts_mech1
# counts_enz1 <- counts_enz1[493:498, 1:3]
# counts_enz1
# counts_mech2 <- counts_mech2[493:498, 1:3]
# counts_mech2
# counts_enz2 <- counts_enz2[493:498, 1:3]
# counts_enz2
#######################################
# Flaten counts (sum all cells for a gene) + combine matrices
mech1 <- as.data.frame(rowSums(counts_mech1))
enz1 <- as.data.frame(rowSums(counts_enz1)) 
mech2 <- as.data.frame(rowSums(counts_mech2)) 
enz2 <- as.data.frame(rowSums(counts_enz2)) 


########################
# Read in bulk samples and create a unified count matrix
########################
# Bulk
b1 <- read.table(file=file_bulk1, sep='\t', header=TRUE)
b2 <- read.table(file=file_bulk2, sep='\t', header=TRUE)
#combined bulk matrix
flat_matrix_BULK <- merge(b1, b2, by ='Geneid', all=TRUE ) #merge by rowname
colnames(flat_matrix_BULK) <- c('ensID',bulk_base_names)

# GT
gt1 <- read.table(file=file_bulkGT1, sep='\t', header=TRUE)
gt2 <- read.table(file=file_bulkGT2, sep='\t', header=TRUE)
#combined bulk matrix
flat_matrix_BULK_GT <- merge(gt1, gt2, by ='Geneid', all=TRUE ) #merge by rowname
colnames(flat_matrix_BULK_GT) <- c('ensID',bulkgt_base_names)

# Matched normal
norm1 <- read.table(file=file_bulkNORM1, sep='\t', header=TRUE)
norm2 <- read.table(file=file_bulkNORM2, sep='\t', header=TRUE)
#combined bulk matrix
flat_matrix_BULK_norm <- merge(norm1, norm2, by ='Geneid', all=TRUE ) #merge by rowname
colnames(flat_matrix_BULK_norm) <- c('ensID',bulknorm_base_names)
```



# Part 1: bulk GT vs bulk B 

### A. Condition only affects

An analysis we can't run is split by patient for bulk gt vs bulk b - because we must have at least a sample size of 2 for each condition. `The design matrix has the same number of samples and coefficients to fit, so estimation of dispersion is not possible.`

If we still want to investigate this trend we can only compare bulk gt vs bulk b when we group together patients and this analysis is included in this section.

```{r}
############################
# Create count matrices for DGE
############################
# Combine matrices
flat_matrix_ctrlBULK <- merge(flat_matrix_BULK, flat_matrix_BULK_GT, by='ensID', all=TRUE)
# Replace NA --> 0 
flat_matrix_ctrlBULK[is.na(flat_matrix_ctrlBULK)]=0
# Now that have combined properly fix so that rowname are genes
flat_matrix_ctrlBULK <- column_to_rownames(flat_matrix_ctrlBULK, var='ensID')

############################
# Create meta data table of which samples are which treatment - needed for Deseq2
#############################
# Grab samples in order
samples <- colnames(flat_matrix_ctrlBULK)

# Create condition vector and type
condition <- c('bulk','bulk', 'gt', 'gt')
meta_data <- as.data.frame(cbind(condition) )
rownames(meta_data) <- samples
meta_data
# # Remove gene rows that are lowly expr in samples - filter low counts <10 ct
# flat_matrix_ctrlBULK <- flat_matrix_ctrlBULK[rowSums(flat_matrix_ctrlBULK > 10) != 0, ]

#####################
# DGE
#####################
dds_obj_ctrlBULK <- DESeqDataSetFromMatrix(countData = flat_matrix_ctrlBULK, colData = meta_data, design = ~ condition)
# DGE based on neg binom distr
dds_ctrlBULK <- DESeq(dds_obj_ctrlBULK, test ='Wald')
res_ctrlBULK <- results(dds_ctrlBULK, alpha=THRES, pAdjustMethod = 'BH')
( mcols(res_ctrlBULK)$description ) # info

###############
# Examine results
###############
#MA plot
print('MA plot of non-transformed counts')
plotMA(res_ctrlBULK, ylim=c(-5,5), alpha = THRES)

# "Table" of top diff expr
print('Most differ expr genes of normalized counts')
topgenes <- rownames(res_ctrlBULK)[res_ctrlBULK$padj <= sort(res_ctrlBULK$padj)[75] &!is.na(res_ctrlBULK$padj)] #for downstream anlaysis
res_ctrlBULK[order(res_ctrlBULK$padj)[0:75],] #top genes table
```

We see there is no significant impact between bulk gt and bulk b2. 

# Part 2A. Establish that single cell protocol does impact gexp

To accomblish this we will see if there is a difference in gene expression between bulk (combined GT,B) vs sc (combined enz and mech)

using wald test (because only 2 factor levels). currently not incorporating batch in linear model

```{r, message=FALSE, echo=FALSE}
############################
# Create count matrices for DGE
############################
# Combine mech samples
flat_matrix_mech <- merge(mech1, mech2, by =0, all=TRUE) #merge by rowname
colnames(flat_matrix_mech) <- c("ensID",scmech_base_names)
# Combine enz samples
flat_matrix_enz <- merge(enz1, enz2, by =0, all=TRUE) #merge by rowname
colnames(flat_matrix_enz) <- c("ensID",scenz_base_names)
# Combine mech and enz samples
flat_matrix_ALL <- merge(flat_matrix_mech, flat_matrix_enz, by ='ensID', all=TRUE) #merge by rowname
# add in bulk
flat_matrix_ALL <- merge(flat_matrix_BULK, flat_matrix_ALL, by='ensID', all=TRUE)
# add in bulk gt
flat_matrix_ALL <- merge(flat_matrix_BULK_GT, flat_matrix_ALL, by='ensID', all=TRUE)
# Replace NA --> 0 
flat_matrix_ALL[is.na(flat_matrix_ALL)]=0
# Now that have combined properly fix so that rowname are genes
flat_matrix_ALL <- column_to_rownames(flat_matrix_ALL, var='ensID')


############################
# Create meta data table of which samples are which treatment - needed for Deseq2
#############################
# Grab samples in order
samples <- colnames(flat_matrix_ALL)
# Create condition vector and type
condition <- c('bulk','bulk','bulk','bulk', 'sc', 'sc', 'sc', 'sc') 
meta_data <- as.data.frame(cbind(condition))
rownames(meta_data) <- samples
meta_data
#### consider removing
# # Remove gene rows that are lowly expr in samples - filter low counts <10 ct
# flat_matrix_ctrlBULK <- flat_matrix_ctrlBULK[rowSums(flat_matrix_ctrlBULK > 10) != 0, ]
```

```{r, message=FALSE, echo=FALSE}
#####################
# DGE
#####################

dds_obj_ALL <- DESeqDataSetFromMatrix(countData = flat_matrix_ALL, colData = meta_data, design = ~ condition)

# DGE based on neg binom distr
dds_ALL <- DESeq(dds_obj_ALL, test ='Wald')
res_ALL <- results(dds_ALL, alpha=THRES, pAdjustMethod = 'BH', test='Wald')
( mcols(res_ALL)$description ) # info
```
```{r, echo=FALSE}
#MA plot
print('MA plot of normalized counts')
plotMA(res_ALL, ylim=c(-5,5), alpha = THRES)

# "Table" of top diff expr
print('Most differ expr genes of normalized counts')
topgenes <- rownames(res_ALL)[res_ALL$padj <= sort(res_ALL$padj)[75] &!is.na(res_ALL$padj)]#downstream anlaysis
res_ALL[order(res_ALL$padj)[0:75],] #top genes table
```

```{r}
print('Now lets look at the log10(counts) of the data')
```
```{r, echo=FALSE}
##############
# Visualization
##############
# 1A. Extract normalized count matrix used by DESeq
    # estimating size factors is auto in DESEQ() but 
    # we have to do this manually to extract the normalized count matrix it produces
    # Normalization method= deseq median or ratios method
dds <- DESeqDataSetFromMatrix(countData = flat_matrix_ALL, colData = meta_data, design = ~condition)
dds<- estimateSizeFactors(dds)
('size factors applied for norm were:')
(sizeFactors(dds)) #good for record keeping
normalized_counts_tibble <- counts(dds, normalized=TRUE)
#write.table(normalized_counts_tibble,file="/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/03_compare-bulk_scrna/scrna2bulk/normalized_counts--ALL.txt", sep="\t", quote=F, col.names=NA)
```

Now let's take a closer look at the top 75 diff expr genes

```{r}
# 1B. get subset of top genes to look at, extract this info from DESeq then subset normalized count matrix
# Reformat
m_tibble <- meta_data %>% 
  rownames_to_column(var="samplename") %>% 
  as_tibble()
normalized_counts <- normalized_counts_tibble %>% 
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# Extract
top75_norm <- normalized_counts %>%
        filter(gene %in% topgenes)


# 2. Plot signficant DE genes
# Gathering the columns to have normalized counts to a single column (aka create a col sample name)
gathered <- top75_norm %>%
  gather(colnames(top75_norm)[2:7], key = "samplename", value = "normalized_counts") #2:7 bc headers are gene,then all 6 samples
# melt so that plot will be colored by condition
gathered <- inner_join(m_tibble, gathered,by="samplename")
# Let's apply log 10 to normalized counts + replace inf to NA from log10 transform
newcol <-log10(gathered$normalized_counts)
is.na(newcol) <- sapply(newcol, is.infinite)
gathered$log10_normalized_counts <- newcol
ggplot(gathered) +
        geom_point(aes(x = gene, y = log10_normalized_counts, color = condition)) +
        xlab("Genes (Ranked)") +
        ylab("Normalized Counts") +
        ggtitle("Top 75 Significant DE Genes") +
        theme_bw() +
	theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
	theme(plot.title = element_text(hjust = 0.5))

# 3. Heatmap
#plot only signficant genes and display their normalized expression
print('Heatmap of TOP GENE log10 normalized counts')
# Reformat 
annotation <- m_tibble %>%data.frame(row.names = "samplename")
mat<- top75_norm %>% column_to_rownames(var = "gene")
# Let's apply log 10 to normalized counts + replace inf to NA from log10 transform
mat <- log10(mat)
is.na(mat) <- sapply(mat, is.infinite)
pheatmap(mat, 
         cluster_rows = T,
         cluster_cols = T,
         show_rownames = T,
         border_color = NA,
         annotation = annotation, 
         fontsize = 5)
```


```{r}
print('Now lets look at all genes not just top 75')
# 3. Heatmap
#plot only signficant genes and display their normalized expression
print('Heatmap of ALL log 10 normalized counts')

# Reformat 
annotation <- m_tibble %>%data.frame(row.names = "samplename")
mat<- normalized_counts %>% column_to_rownames(var = "gene")
# Let's apply log 10 to normalized counts + replace inf to 0 from log10 transform
mat <- log10(mat)
# mat <- mat[1:20000,]
is.na(mat) <- sapply(mat, is.infinite)
mat <- na.omit(mat) 
pheatmap(mat, 
         cluster_rows = T,
         cluster_cols = T,
         show_rownames = F,
         border_color = NA,
         annotation = annotation, 
         fontsize = 5)
```

```
### note underlying assumption is that we are just removing genes where at least one sample has inf val after log transform. this needs to be patched to handle instead of remove, but for the purposes of showing global batch effect it won't impact it for WCS. BUT we must address for final analysis
```








# Part 2B. Comparing scRNA-mech pseudo, scRNA-enz pseudo, and bulk - normalized

using LRT (not wald) bc factor level>2. not yet incorporate batch in linear model


```{r, message=FALSE}
############################
# Create count matrices for DGE
############################
# Combine mech samples
flat_matrix_mech <- merge(mech1, mech2, by =0, all=TRUE) #merge by rowname
colnames(flat_matrix_mech) <- c("ensID",scmech_base_names)
# Combine enz samples
flat_matrix_enz <- merge(enz1, enz2, by =0, all=TRUE) #merge by rowname
colnames(flat_matrix_enz) <- c("ensID",scenz_base_names)
# Combine mech and enz samples
flat_matrix_ALL <- merge(flat_matrix_mech, flat_matrix_enz, by ='ensID', all=TRUE) #merge by rowname
# Combine mech, enz, bulk
flat_matrix_ALL <- merge(flat_matrix_BULK, flat_matrix_ALL, by='ensID', all=TRUE)
# Replace NA --> 0 
flat_matrix_ALL[is.na(flat_matrix_ALL)]=0
# Now that have combined properly fix so that rowname are genes
flat_matrix_ALL <- column_to_rownames(flat_matrix_ALL, var='ensID')


############################
# Create meta data table of which samples are which treatment - needed for Deseq2
#############################
# Grab samples in order
samples <- colnames(flat_matrix_ALL)
# Create condition vector and type
condition <- c('bulk','bulk', 'mech', 'mech', 'enz', 'enz')
meta_data <- as.data.frame(condition)
rownames(meta_data) <- samples
meta_data


#### consider removing
# # Remove gene rows that are lowly expr in samples - filter low counts <10 ct
# flat_matrix_ctrlBULK <- flat_matrix_ctrlBULK[rowSums(flat_matrix_ctrlBULK > 10) != 0, ]
```

```{r, message=FALSE}
#####################
# DGE
#####################
dds_obj_ALL <- DESeqDataSetFromMatrix(countData = flat_matrix_ALL, colData = meta_data, design = ~condition)

# DGE based on neg binom distr
dds_ALL <- DESeq(dds_obj_ALL, test ='LRT', reduced=~1)
res_ALL <- results(dds_ALL, alpha=THRES, pAdjustMethod = 'BH')
( mcols(res_ALL)$description ) # info

#MA plot
print('MA plot of normalized counts')
plotMA(res_ALL, ylim=c(-5,5), alpha = THRES)

# "Table" of top diff expr
print('Most differ expr genes of normalized counts')
topgenes <- rownames(res_ALL)[res_ALL$padj <= sort(res_ALL$padj)[75] &!is.na(res_ALL$padj)]
#downstream anlaysis
res_ALL[order(res_ALL$padj)[0:75],] #top genes table
```

Now let's look at the log10(counts) of the data

```{r}
##############
# Visualization
##############
# First, Extract normalized count matrix used by DESeq
    # estimating size factors is auto in DESEQ() but 
    # we have to do this manually to extract the normalized count matrix it produces
    # Normalization method= deseq median or ratios method
dds <- DESeqDataSetFromMatrix(countData = flat_matrix_ALL, colData = meta_data, design = ~condition)
dds<- estimateSizeFactors(dds)
('size factors applied for norm were:')
(sizeFactors(dds)) #good for record keeping
normalized_counts_tibble <- counts(dds, normalized=TRUE)
#write.table(normalized_counts_tibble,file="/Users/leejor/Ellrott_Lab/04_CSBC/cell-dissociation/data/03_compare-bulk_scrna/scrna2bulk/normalized_counts--ALL.txt", sep="\t", quote=F, col.names=NA)
```

Now let's take a closer look at the top 75 diff expr genes

```{r}
# Get subset of top genes to look at, extract this info from DESeq then subset normalized count matrix
# Reformat
m_tibble <- meta_data %>% 
  rownames_to_column(var="samplename") %>% 
  as_tibble()
normalized_counts <- normalized_counts_tibble %>% 
  data.frame() %>%
  rownames_to_column(var="gene") %>% 
  as_tibble()
# Extract
top75_norm <- normalized_counts %>%
        filter(gene %in% topgenes)


# 2. Plot signficant DE genes
# Gathering the columns to have normalized counts to a single column (aka create a col sample name)
gathered <- top75_norm %>%
  gather(colnames(top75_norm)[2:7], key = "samplename", value = "normalized_counts") #2:7 bc headers are gene,then all 6 samples
# melt so that plot will be colored by condition
gathered <- inner_join(m_tibble, gathered,by="samplename")
# Let's apply log 10 to normalized counts + replace inf to NA from log10 transform
newcol <-log10(gathered$normalized_counts)
is.na(newcol) <- sapply(newcol, is.infinite)
gathered$log10_normalized_counts <- newcol
#plot
ggplot(gathered) +
        geom_point(aes(x = gene, y = log10_normalized_counts, color = condition)) +
        xlab("Genes (Rank)") +
        ylab("Log 10 Normalized Counts") +
        ggtitle("Top 75 Significant DE Genes") +
        theme_bw() +
	theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
	theme(plot.title = element_text(hjust = 0.5))

# 3. Heatmap
#plot only signficant genes and display their normalized expression
print('Heatmap of TOP GENE log 10 normalized counts')
# Reformat 
annotation <- m_tibble %>%data.frame(row.names = "samplename")
mat<- top75_norm %>% column_to_rownames(var = "gene")
# Let's apply log 10 to normalized counts + replace inf to NA from log10 transform
mat <- log10(mat)
is.na(mat) <- sapply(mat, is.infinite)
pheatmap(mat, 
         cluster_rows = T,
         cluster_cols = T,
         show_rownames = T,
         border_color = NA,
         annotation = annotation, 
         fontsize = 5)
```

```{r}
print('Now lets look at all genes not just top 75')
# 3. Heatmap
#plot only signficant genes and display their normalized expression
print('Heatmap of ALL log 10 normalized counts')
# Reformat 
annotation <- m_tibble %>%data.frame(row.names = "samplename")
mat<- normalized_counts %>% column_to_rownames(var = "gene")
# Let's apply log 10 to normalized counts + replace inf to 0 from log10 transform
mat <- log10(mat)
# mat <- mat[1:20000,]
is.na(mat) <- sapply(mat, is.infinite)
mat <- na.omit(mat) 
pheatmap(mat, 
         cluster_rows = T,
         cluster_cols = T,
         show_rownames = F,
         border_color = NA,
         annotation = annotation, 
         fontsize = 5)
```
```
### note underlying assumption is that we are just removing genes where at least one sample has inf val after log transform. this needs to be patched to handle instead of remove, but for the purposes of showing global batch effect it won't impact it for WCS. BUT we must address for final analysis
```

# Part 3: Tumor vs matched normal

Comparing differences between bulk and N (excluding GT from this analysis)

```{r}
############################
# Create count matrices for DGE
############################
# Combine matrices
flat_matrix_ctrlBULK <- merge(flat_matrix_BULK, flat_matrix_BULK_norm, by='ensID', all=TRUE)
# Replace NA --> 0 
flat_matrix_ctrlBULK[is.na(flat_matrix_ctrlBULK)]=0
# Now that have combined properly fix so that rowname are genes
flat_matrix_ctrlBULK <- column_to_rownames(flat_matrix_ctrlBULK, var='ensID')

############################
# Create meta data table of which samples are which treatment - needed for Deseq2
#############################
# Grab samples in order
samples <- colnames(flat_matrix_ctrlBULK)

# Create condition vector and type
condition <- c('bulk','bulk', 'norm', 'norm')
meta_data <- as.data.frame(cbind(condition) )
rownames(meta_data) <- samples
meta_data

# # Remove gene rows that are lowly expr in samples - filter low counts <10 ct
# flat_matrix_ctrlBULK <- flat_matrix_ctrlBULK[rowSums(flat_matrix_ctrlBULK > 10) != 0, ]
```

**Code below throws an error, will need to revisit this to fix**

`Error in estimateSizeFactorsForMatrix(counts(object), locfunc = locfunc, : every gene contains at least one zero, cannot compute log geometric means`

```
#####################
# DGE
#####################
dds_obj_ctrlBULK <- DESeqDataSetFromMatrix(countData = flat_matrix_ctrlBULK, colData = meta_data, design = ~ condition)
# DGE based on neg binom distr
dds_ctrlBULK <- DESeq(dds_obj_ctrlBULK, test ='Wald')
```






